File: 2025/02/prog_76502_괄호회전하기.py
# 올바른 괄호 문자열이 만들어지는 경우의 수를 출력하기 
from collections import deque
def solution(s):
    answer = 0
    q = deque(list(s))
    
    open_b = ['[', '{', '(']
    close_b  = [']', '}', ')']
    
    def is_perfect_bracket(bracket):
        stack = []
        
        for bb in bracket:
            if bb in open_b:
                stack.append((bb, open_b.index(bb)))
            else:
                if not stack:
                    return False
                top_b, top_n = stack.pop()
                if close_b.index(bb) != top_n:
                    return False
                
        if stack:
            return False
        
        return True
                
    
    for x in range(0, len(s)):
        # 왼쪽으로 1칸 회전 
        front = q.popleft()
        q.append(front)
        
        # 회전한 결과 올바른 괄호인가? 
        if is_perfect_bracket(q):
            answer += 1
    
    return answer

File: 2025/07/prog_76502_괄호회전하기.js
function isRight(arr) {
    let stack = []
    const openB = {'(': 0, '{': 1, '[': 2, a: 3}
    const closeB = {')': 0, '}': 1, ']': 2}
    
    for (let a of arr) {
        if (a in openB) {
            stack.push(a)
        } else {
            if (stack.length === 0) {
                return false
            }
            
            const open = stack.pop()
            if (!(open in openB)) {
                return false
            }

            const openI = openB[open]
            const closeI = closeB[a]           
            
            if (openI !== closeI) {
                return false
            }
        }
    }
    
    if (stack.length !== 0) {
        return false
    }
    
    return true
}


function solution(s) {
    var answer = 0;
    let arr = s.split("")
    
    // 총 몇 번 쉬프트? 
    for (let i = 0; i < arr.length; i++) {
        // 쉬프트하기 
        const front = arr.shift()
        arr.push(front)
 
        if (isRight(arr)) answer += 1
    }
    
    return answer;
}